docs/proposals/operator-custom-crd-download-url/README.md

```diff
@@ -0,0 +1,79 @@
+is based on the intended version of that instance, not the release version of the operator.
+
+## Proposal
+
+Currently, when provisioning a new Karmada instance, as a prerequisite, the Karmada operator will ensure that required CRDs are downloaded. By default, the CRDs tarball is
+downloaded from a well-known location on GitHub based on the release version of the operator. As aforementioned, this is based on the assumption that the version of the Karmada instance 
+is identical to that of the operator's release version, which may not always be true. When the version of the Karmada control plane components is not specified, the default behavior is to assume
+those components should have the same version as the operator's release version. In that scenario, the assumption is justified and holds true. However, when creating a new Karmada instance, it's possible to explicitly
+specify the version of the control plane components. In that scenario, this assumption breaks, and we encounter the risk of the operator installing the wrong CRDs for that Karmada instance.
+
+The default behavior is sensible and will be retained. However, it should not be the behavior when the intent is to create a new Karmada instance with control plane components of a different version than that of the operator's release version. 
+To support that scenario, we require the ability to explicitly set, as part of the Karmada resource, the URL that should be used to download the CRDs tarball for that Karmada instance.
+
+In addition to making sure required CRDs are downloaded, the Karmada operator also employs a caching strategy to avoid re-downloading CRDs it has already downloaded. It's a cache on the local file system.
+Downloaded CRDs are stored at a well-known location on the local file system based the release version of the Karmada operator. When provisioning a new Karmada instance, before attempting to download the required CRDs, the operator will fist check if
+they already exist in the cache. To support custom CRD download URLs, we require this caching strategy be extended to be more generic.
+
+## Design Details
+
+To support downloading CRDs from a custom URL, we need to add a new field called `CRDRemoteURL` to the Karmada resource spec:
+```go
+type KarmadaSpec struct {
+	HostCluster *HostCluster `json:"hostCluster,omitempty"`
+
+	PrivateRegistry *ImageRegistry `json:"privateRegistry,omitempty"`
+	
+	Components *KarmadaComponents `json:"components,omitempty"`
+	
+	FeatureGates map[string]bool `json:"featureGates,omitempty"`
+
+	// Custom URL that should be used to download required CRDs for this Karmada instance.
+	// Omit if the intent is to provision a new Karmada instance with control plane components of the same version as the Karmada operator's release version.
+	// Otherwise, it should be explicitly set to ensure the right version of the CRDs is installed for this instance.
+	// +optional
+	CRDRemoteURL string `json:"crdRemoteURL,omitempty"`
```

Comment:
In fact, within our company, we have copied the crd file into the image of karmada-operator.

So what I want to ask is, if my crd file is stored in the image, how should I fill in the crdRemoteURL?

Reply:
> we have copied the crd file into the image of karmada-operator.

How to do that? What I can imagine there are two ways to do that:
1. Using a private karmada-operator image, the tarball can be put into the image
2. Mount the tarball into the pod, by a ConfigMap when deploying the operator

The tarball is located in the cache, so karmada-opeator doesn't need to download the tarball from GitHub.

Reply:
So what I mean is, should `crdRemoteURL` support a format like `file:///var/loca/crds`?

Reply:
This is possible and can be supported (e.g., the cache entry can just be a symbolic link the referenced location on the local filesystem), but can you please elaborate more on why that strategy would be preferred to downloading the CRDs from a remote location? How would you manually manage storage for multiple versions of the CRDs? Would that not require more work and maintenance on your part?

Reply:
> but can you please elaborate more on why that strategy would be preferred to downloading the CRDs from a remote location?


Simply put, we don't want our end users to download from remote http/https servers. Because that requires our end users to do some extra work, such as preparing additional http/https servers, etc. We would rather have the karmada-operator read the crd file from the karmada-operator image.

In this way, users only need to use our provided `Web-UI` function to deploy a brand new karmada instance with `one click`. There's no need for extra attention on where the crd is stored.